Notification on Comment Made

tinfernn opened the pull request:
# Feature Addition

## Summary

When a comment is made on an incident, an email, SMS, and database notification will be sent to all admins and to the assigned supervisor (only if a supervisor is assigned).

## Rationale

This will notify all admins and the assigned supervisor (if one is assigned) to comments made on an incident, which will keep everyone up to date on any progress or changes made to an incident. This allows for admins and supervisors to respond quickly to comments made as well.

## Design Documentation

N/A

## Changes

Changes the react() method in CommentCreated.php to send the necessary notifications.

## Impact

N/A

## Testing

* The following tests have been added to CommentCreatedTest.php:
    * Notifies admin on comment made
    * New comment notifies supervisor when supervisor is set
    * New comment does not notify supervisor when supervisor is not set
    * New comment notifies both admin and set supervisor
* The following tests have been added to IncidentAggregateRootTest.php:
    * Comment notifies admin team
    * Comment notifies supervisor when supervisor is set
    * Comment does not notify supervisor when supervisor is not set
    * Comment notifies admin team and supervisor when supervisor is set
*  The following tests have been added to AddCommentTest.php:
    * Comment notifies admins on creation
    * Comment notifies supervisor when supervisor is set
    * Comment does not notify supervisor when not set

All tests are passing mentioned above are passing.

## Screenshots/Video

N/A

## Checklist

- [x] Code follows the project's coding standards

- [x] Unit tests covering the new feature have been added

- [x] All existing tests pass

- [ ] The documentation has been updated to reflect the new feature

## Additional Notes

N/A

tinfernn committed:
Created CommentMade.php to send SMS, email, and database notification. Edited the react() method in CommentCreated.php to send notifications to all admins and to the assigned supervisor (only if one is assigned to the incident).

Created tests in AddCommentTest.php, CommentCreatedTest.php, and IncidentAggregateRootTest.php

tinfernn committed:
Merge branch 'staging' into notification-on-comment-made

welshy557 commented on app/Notifications/Comment/CommentMade.php:
```diff
@@ -0,0 +1,62 @@
+    public function toArray(object $notifiable): array
+    {
+        return [
+            'message' => "A new comment has been made.",
```
- Make the message say: "{name} commented: {comment text}". Put it in a variable and pass it to toArray and toVonage messages to it's consistent.
- Pass URL to toArray method. See other notifications for example.

welshy557 commented on app/Notifications/Comment/CommentMade.php:
```diff
@@ -0,0 +1,62 @@
+    {
+        return (new MailMessage)
+            ->subject('Comment Created')
+            ->markdown('mail.comment-made');
```
Pass in the comment text and commenter name and render it in the email so the email has some context.

welshy557 commented on resources/views/mail/comment-made.blade.php:
```diff
@@ -0,0 +1,10 @@
+<x-mail::message>
```
Update with passed in commenter name and comment text

tinfernn committed:
Edited CommentMade.php to address the following for toArray:
Make the message say: "{name} commented: {comment text}".
-Put it in a variable and pass it to toArray and toVonage messages to it's consistent.
-Pass URL to toArray method. See other notifications for example.

Edited CommentMade.php to address the following for toMail:
-Pass in the comment text and commenter name and render it in the email so the email has some context.

Edited comment-made.blade.php to address the following:
-Update with passed in commenter name and comment text

Edited react() in CommentCreated.php to allow for the changes mentioned above for CommentMade.php

OsmiumCoder commented on app/Notifications/Comment/CommentMade.php:
```diff
@@ -0,0 +1,81 @@
+use Illuminate\Notifications\Notification;
+use Illuminate\Notifications\Messages\VonageMessage;
+
+class CommentMade extends Notification
```
Change name to CommentAdded, class name and file name

OsmiumCoder commented on app/StorableEvents/Comment/CommentCreated.php:
```diff
@@ -29,4 +32,20 @@ public function handle()
+    {
+        $commentable = $this->commentable_type::find($this->commentable_id);
+        $commenter = User::find($this->metaData['user_id'] ?? null);
+        $commenterName = $commenter ? "{$commenter->first_name} {$commenter->last_name}" : null;
```
user only has name so just do $commenter->name

OsmiumCoder replied:
actually just remove this and just have commenter, change commenter to user and pass user to the notification

OsmiumCoder commented on app/Notifications/Comment/CommentMade.php:
```diff
@@ -0,0 +1,81 @@
+     */
+    public function __construct(
+        public string $comment,
+        public ?string $commenterName,
```
as in other comment just pass public ?User $user in here

OsmiumCoder commented on app/Notifications/Comment/CommentMade.php:
```diff
@@ -0,0 +1,81 @@
+            ->subject('Comment Created')
+            ->line($this->message)
+            ->markdown('mail.comment-made', [
+                'recipient' => $notifiable->name,
```
don't think we need recipient

OsmiumCoder commented on resources/views/mail/comment-made.blade.php:
```diff
@@ -0,0 +1,10 @@
+<x-mail::message>
+    # Comment Created
+
+    Dear {{ $recipient }},
```
i think we are fine without Dear here just who made comment and what

OsmiumCoder commented on tests/Feature/Incident/AddCommentTest.php:
```diff
@@ -99,4 +101,63 @@ public function test_creates_comment_and_attaches_to_model()
         $this->assertEquals(CommentType::NOTE, $comment->type);
     }
+
+    public function test_comment_notifies_admins_on_creation()
```
name of this tests says we are testing to see if admins receive the notification, but no users exist in this test and we are asserting nothing was sent

OsmiumCoder commented on tests/Feature/Incident/AddCommentTest.php:
```diff
@@ -99,4 +101,63 @@ public function test_creates_comment_and_attaches_to_model()
+        Notification::assertSentTo($supervisor, CommentMade::class);
+    }
+
+    public function test_comment_does_not_notify_supervisor_when_not_set()
```
make a supervisor and verify it was not sent to them, also have admins and verify it was sent to them

OsmiumCoder commented on tests/Feature/Incident/AddCommentTest.php:
```diff
@@ -99,4 +101,63 @@ public function test_creates_comment_and_attaches_to_model()
+        Notification::assertNothingSent();
+    }
+
+    public function test_comment_notifies_supervisor_when_supervisor_is_set()
```
also add an assertion to check that it was sent to admins

OsmiumCoder commented on tests/Unit/Aggregates/IncidentAggregateRootTest.php:
```diff
@@ -736,4 +737,107 @@ public function test_create_incident_notifies_admin_team(): void
+        $supervisor = User::factory()->create()->syncRoles('supervisor');
+
+        $incident = Incident::factory()->create([
+            'supervisor_id' => null,
```
test name indicates supervisor is set in this test but it isn't, it should be as next test it is not set

OsmiumCoder commented on tests/Unit/StoreableEvents/Comment/CommentCreatedTest.php:
```diff
@@ -71,4 +73,108 @@ public function test_stores_comment()
+
+        $event->setMetaData([
+            'user_id' => $user->id,
+            'supervisor' => null,
```
don't need this line

OsmiumCoder committed:
resolved merge conflicts

OsmiumCoder commented:
also in all the tests you added add another test to verify that the to array is correct for database store, for a good example of this see IncidentReviewRequested event, incident aggregate root, and the feature/incident/status tests for how this was tested

tinfernn committed:
Merge branch 'refs/heads/staging' into notification-on-comment-made

# Conflicts:
#	tests/Unit/Aggregates/IncidentAggregateRootTest.php

tinfernn committed:
Fixed tests.

Refactored CommentMade to CommentAdded and changed constructor + made edits in some of the methods.

Edited comment-made.blade.php

tinfernn committed:
Changed $user to not be nullable.

Fixed tests.

tinfernn committed:
Edit to react()

OsmiumCoder committed:
resolved merge conflicts

OsmiumCoder committed:
lint
